--- setup.py.orig	2015-06-19 09:17:25.000000000 +0900
+++ setup.py	2017-07-05 00:24:23.000000000 +0900
@@ -2,10 +2,18 @@
 from distutils.extension import Extension
 from Cython.Distutils import build_ext
 import os
-os.environ['CFLAGS'] = "-I/usr/include/libusb-1.0"
-os.environ['LDFLAGS'] = "-L/usr/lib/i386-linux-gnu -lusb-1.0 -ludev -lrt"
+os.environ['CFLAGS'] = "-framework IOKit -framework CoreFoundation"
+os.environ['LDFLAGS'] = ""
 
 setup(
+    name="hidapi",
+    version="0.7.99",
+    author="Gary Bishop",
+    author_email="gb@cs.unc.edu",
+    description=("A cython interface to the hidapi from https://github.com/signal11/hidapi with the C modification for windows from http://code.google.com/p/picusb/downloads/detail?name=hidapi_git_mingw_7e93a4e068825d227807.zip&can=2&q= so I could build it with mingw on windows."),
+    license="BSD",
+    url="https://github.com/gbishop/cython-hidapi",
+    #packages=['hid'],
     cmdclass = {'build_ext': build_ext},
-    ext_modules = [Extension("hid", ["hid.pyx", "hid-libusb.c"])]
+    ext_modules = [Extension("hid", ["hid.pyx", "hid-mac.c"])]
 )
