--- PyDSTool/parseUtils.py.org	2015-03-31 04:24:39.000000000 +0900
+++ PyDSTool/parseUtils.py	2018-01-15 23:06:39.000000000 +0900
@@ -43,8 +43,8 @@
             'y0', 'y1', 'i0', 'i0e', 'i1', 'i1e', 'k0', 'k0e',
             'k1', 'k1e', 'itj0y0', 'it2j0y0', 'iti0k0', 'it2i0k0',
             'besselpoly', 'jvp', 'yvp', 'kvp', 'ivp', 'h1vp',
-            'h2vp', 'sph_jn', 'sph_yn', 'sph_jnyn', 'sph_in',
-            'sph_kn', 'sph_inkn', 'riccati_jn', 'riccati_yn',
+            'h2vp', 'spherical_jn', 'spherical_yn', 'spherical_in',
+            'spherical_kn', 'riccati_jn', 'riccati_yn',
             'struve', 'modstruve', 'itstruve0', 'it2struve0',
             'itmodstruve0', 'bdtr', 'bdtrc', 'bdtri', 'btdtr',
             'btdtri', 'fdtr', 'fdtrc', 'fdtri', 'gdtr', 'gdtrc',
