--- xgboost/build-python.sh.org	2019-11-28 17:44:42.000000000 +0900
+++ xgboost/build-python.sh	2019-11-28 17:45:33.000000000 +0900
@@ -19,12 +19,15 @@
 if echo "${OSTYPE}" | grep -q "darwin"; then
   LIB_XGBOOST=libxgboost.dylib
   # Use OpenMP-capable compiler if possible
-  if which g++-5; then
-    export CC=gcc-5
-    export CXX=g++-5
-  elif which g++-7; then
-    export CC=gcc-7
-    export CXX=g++-7
+  if which g++-mp-8; then
+    export CC=gcc-mp-8
+    export CXX=g++-mp-8
+  elif which g++-mp-7; then
+    export CC=gcc-mp-7
+    export CXX=g++-mp-7
+  elif which g++-mp-5; then
+    export CC=gcc-mp-5
+    export CXX=g++-mp-5
   elif which clang++; then
     export CC=clang
     export CXX=clang++
