--- pysces/PyscesConfig.py.dist	2017-09-11 21:35:21.000000000 +0900
+++ pysces/PyscesConfig.py	2017-09-11 21:53:23.000000000 +0900
@@ -70,6 +70,8 @@
         "output_dir"   : os.path.join(os.path.expanduser('~'),'Pysces'),
         "silentstart"  : False
         }
+    if os.sys.platform == 'darwin':
+        __DefaultPosix["matplotlib_backend"] = 'MacOSX'
 
 def ReadConfig(file_path, config={}):
     """
