--- setup.py.dist	2017-09-11 22:51:58.000000000 +0900
+++ setup.py	2017-09-11 22:53:14.000000000 +0900
@@ -137,7 +137,10 @@
     #add configuration data
     cp.add_section('PyscesConfig')
     cp.set('PyscesConfig','matplotlib', True)
-    cp.set('PyscesConfig','matplotlib_backend', 'TkAgg')
+    if os.sys.platform == 'darwin':
+        cp.set('PyscesConfig','matplotlib_backend', 'MacOSX')
+    else:
+        cp.set('PyscesConfig','matplotlib_backend', 'TkAgg')
     cp.set('PyscesConfig','gnuplot', False)
     # Built in modules
     cp.add_section('PyscesModules')
